### PR TITLE
support pluggable preparation hooks through a plugin system

### PR DESCRIPTION
**Pull request by steved**
This is really my opportunity to sneak YAL (yet another logger).

Domino has some custom logic currently necessary to ensure that images created have the correct files in place before the context is finalized. To that end, this is an initial use for a plugin system based around [go-plugin](https://github.com/hashicorp/go-plugin) that can load arbitrary third party binaries, communicate over RPC, and allow custom logic during the "preparation" phase (after the context is created, before the build starts).

TODOs

- [x] README / documentation on how to create a plugin
- [ ] ~~Support a configurable log level for the go-plugin logs~~ address this in a follow-up
- [x] Unit tests, specifically for: rpcServer and LoadPlugins I think (suggestions welcome).

### CODE REVIEW

File: plugins/preparer/client.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"errors"
+	"net/rpc"
+)
+
+type rpcClient struct {
+	client *rpc.Client
+}
+
+var _ Preparer = &rpcClient{}
```
Comment (sonnysideup) [inline plugins/preparer/client.go:12]: What's the function of this initialization?

File: plugins/preparer/plugin.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	Preparer
+}
+
+func (p *Plugin) Server(*plugin.MuxBroker) (interface{}, error) {
+	return &rpcServer{p.Preparer}, nil
+}
+
+func (p *Plugin) Client(b *plugin.MuxBroker, c *rpc.Client) (interface{}, error) {
+	return &rpcClient{client: c}, nil
+}
```
[the remark this answers is not in the document]
Comment (steved) [reply plugins/preparer/plugin.go:19]: If I understand it correctly, it's just a side-channel for data passing:
https://github.com/hashicorp/go-plugin/blob/master/mux_broker.go#L15-L28

File: plugins/preparer/client.go (hunk 1 above)
Comment (steved) [reply plugins/preparer/client.go:12]: Ensures that `rpcClient` adheres correctly to the `Preparer` interface. Not 100% sure it's necessary, but I don't see a place where that check is made implicitly (through arguments or return values) at the moment.

File: cmd/root.go [hunk 3]
```diff
@@ -59,4 +61,5 @@ func init() {
 	rootCmd.Flags().StringVar(&messageBroker, "message-broker", "", fmt.Sprintf("Publish resource state changes to a message broker (supported values: %v)", message.SupportedBrokers))
 	rootCmd.Flags().StringVar(&amqpURI, "amqp-uri", "", "AMQP broker connection URI")
 	rootCmd.Flags().StringVar(&amqpQueue, "amqp-queue", "", "AMQP broker queue name")
+	rootCmd.Flags().StringVar(&preparerPluginsPath, "preparer-plugins-path", "", "Path to specific preparer plugins or directory to load them from")
```
Comment (sonnysideup) [inline cmd/root.go:64]: @steved 

I wonder if we want to use something like `XDG_DATA_HOME/plugins` as a default here? We could then just install plugins into the default path and have them picked up. Thoughts?

Comment (steved) [reply cmd/root.go]: Yep, that makes sense

File: plugins/preparer/server_test.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	"reflect"
+	"testing"
+
+	"github.com/pkg/errors"
+)
+
+type testPreparer struct {
+	err               error
+	calledContextPath string
+	calledPluginData  map[string]string
```
Comment (sonnysideup) [inline plugins/preparer/server_test.go:13]: Do you want to add `calledCleanup bool` so you set that to `true` inside the `Cleanup` func and assert that it was called inside `Test_preparerServer_Cleanup`?

File: plugins/preparer/loader.go [hunk 5]
```diff
@@ -5,7 +5,13 @@ import (
 	"path/filepath"
 )
 
+type pluginLoader = func(string) (*Plugin, error)
```
Comment (sonnysideup) [inline plugins/preparer/loader.go:8]: I might attack this at one level higher so you can exercise the public API:

```go
var DefaultPluginLoader = NewPreparerPlugin

type pluginLoader = func(string) (*Plugin, error)

func LoadPlugins(preparerPluginsPath string) (preparerPlugins []*Plugin, err error) {
	return loadPlugins(preparerPluginsPath, DefaultPluginLoader)
}
```

Inside `TestLoadPlugins`:

```go
// test case loop

DefaultPluginLoader = func(path string) (*Plugin, error) {
	return &Plugin{}, nil
}
defer func() {
	DefaultPluginLoader = NewPreparerPlugin
}()

gotPreparerPlugins, err := LoadPlugins(tt.path)

// assertions
```

**Pull request merged**